### PR TITLE
filter duplicated numbers (sms contacts)

### LINKED ISSUE #4191: Duplicate SMS Contacts
I noticed today that I have some contacts twice in my list but only there. Not TS/Signal contacts but those for sending SMS.

The pattern is the same, a number which I stored formated like +49 171 7654321 is visible like this and as +491717654321. I only could see cell phone number affected.

And I can't see a pattern which contacts are affected and which are not.

### PR DESCRIPTION
**Pull request by AndHell**
filter system contacts to remove duplicated numbers

fixes  #4191

### First time contributor checklist

- [x] I have read [how to contribute](https://github.com/WhisperSystems/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [x] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [x] I am following the [Code Style Guidelines](https://github.com/WhisperSystems/Signal-Android/wiki/Code-Style-Guidelines)
- [x] I have tested my contribution on these devices:
 * Alcatel 4034D, Android 6.6
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)
- [x] I have made the choice whether I want the [BitHub reward](https://github.com/WhisperSystems/Signal-Android/wiki/BitHub-Rewards) or not by omitting or adding the word `FREEBIE` in the commit message of my first commit

----------

### Description

Filter duplicated system contacts by running over the curser and check for duplicated numbers after removing everything but 0-9 and +#. 
This removes the duplicated contacts, if another app like WhatsApp adds a new number in a different format to the system contact list.

I'm not sure if this is the right place or way to fix this. If not, what would be a better way?

//FREEBIE

### CODE REVIEW

File: src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java [hunk 1]
```diff
@@ -175,9 +176,37 @@ public ContactsDatabase(Context context) {
...
+
+  private Cursor getFilteredSystemContacts(Cursor cursor, String[] projection) {
+    MatrixCursor filterCursor = new MatrixCursor(projection);
+
+    List<SystemContactInfo> nums = new ArrayList<SystemContactInfo>();
+
+    while (cursor != null && cursor.moveToNext()) {
+      String shortNumber = cursor.getString(2).replaceAll("[^\\d+*#]", "");
+
+      boolean isInList = false;
```
Comment (JonasVautherin) [inline src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java]: I would put lines 193-199 in a `isDuplicate(Cursor cursor)` function, to make the code cleaner. Then, on line 200, I would test `if (isDuplicate(cursor)) { continue; }`

**Comment by moxie0** [comment]
Hmm, it looks to me like this does an N^2 scan over all contacts in the result set and loads them all into memory. I think we'd need something that performs better.

**Comment by JonasVautherin** [comment]
I had started a review 8 days ago, but I apparently forgot to submit. Just did it, sorry for that :-)

**Comment by AndHell** [comment]
> I had started a review 8 days ago, but I apparently forgot to submit. Just did it, sorry for that :-)

Thanks for the feedback and the links.


I took a look at the SQL query to get the numbers and updated to remove the duplicates. This should perform better than my previous version. 
Another benefit is that if a signal contact has stored multiple numbers that are not on signal, these numbers are now shown in the sms contact list. Before the whole contact get blocked with every number.

File: src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java [hunk 2]
```diff
@@ -161,9 +161,11 @@ public ContactsDatabase(Context context) {
...
     }};
 
-    String excludeSelection = ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " NOT IN (" +
-        "SELECT data.contact_id FROM raw_contacts, view_data data WHERE raw_contacts._id = data.raw_contact_id AND " +
-        "data.mimetype = '" + CONTACT_MIMETYPE + "')";
+    String formattedNumber = "REPLACE(REPLACE(REPLACE(REPLACE(data1,' ',''),'-',''),'(',''),')','')";
+    String excludeSelection = "(" + formattedNumber +" NOT IN " +
+            "(SELECT data1 FROM view_data WHERE "+formattedNumber+" = data1) " +
+            "OR "+formattedNumber+" = data1)" +
+            "AND " + formattedNumber + "NOT IN (SELECT "+formattedNumber+" FROM view_data where mimetype = '"+CONTACT_MIMETYPE+"')" ;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java]: sorry, i can't follow, can you explain what is going on here?

Comment (AndHell) [inline src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java]: data1 is the field where the phone number is stored

165-166 numbers that are stored with special chars and have no duplicate without these
167 every number that is stored without special chars
168  exclude every number that is on signal

165-167 together returns every number without the duplicates

**Comment by moxie0** [comment]
in 4.6.0

**Pull request closed without merging**